Kernel/Arch/aarch64/Interrupts.cpp

```diff
@@ -5,27 +5,132 @@
+
+        // TODO: Add these interrupts as a source of entropy for randomness.
+        u8 irq = 0;
+        while (pending_interrupts) {
+            if ((pending_interrupts & 0b1) != 0b1) {
+                irq += 1;
+                pending_interrupts >>= 1;
+                continue;
+            }
+
+            // TODO: Consider removing passing the RegisterState into the handle_interrupt()
+            //       function, since no IRQHandler seems to be using the registers.
+            RegisterState regs {};
+
+            auto* handler = s_interrupt_handlers[irq];
+            VERIFY(handler);
+            handler->increment_invoking_counter();
+            handler->handle_interrupt(regs);
+            handler->eoi();
+
+            irq += 1;
+            pending_interrupts >>= 1;
+        }
+    }
+}
+
+// FIXME: Share the code below with Arch/x86/common/Interrupts.cpp
+//        While refactoring, the interrupt handlers can also be moved into the InterruptManagement class.
+GenericInterruptHandler& get_interrupt_handler(u8 interrupt_number)
+{
+    auto*& handler_slot = s_interrupt_handlers[interrupt_number];
+    VERIFY(handler_slot != nullptr);
+    return *handler_slot;
+}
+
+static void revert_to_unused_handler(u8 interrupt_number)
 {
-    VERIFY_NOT_REACHED();
+    auto handler = new UnhandledInterruptHandler(interrupt_number);
+    handler->register_interrupt_handler();
 }
 
-void register_generic_interrupt_handler(u8, GenericInterruptHandler&)
+void register_generic_interrupt_handler(u8 interrupt_number, GenericInterruptHandler& handler)
 {
-    VERIFY_NOT_REACHED();
+    auto*& handler_slot = s_interrupt_handlers[interrupt_number];
+    if (handler_slot != nullptr) {
+        if (handler_slot->type() == HandlerType::UnhandledInterruptHandler) {
+            if (handler_slot) {
+                auto* unhandled_handler = static_cast<UnhandledInterruptHandler*>(handler_slot);
+                unhandled_handler->unregister_interrupt_handler();
+                delete unhandled_handler;
+            }
+            handler_slot = &handler;
+            return;
+        }
+        if (handler_slot->is_shared_handler() && !handler_slot->is_sharing_with_others()) {
+            VERIFY(handler_slot->type() == HandlerType::SharedIRQHandler);
+            static_cast<SharedIRQHandler*>(handler_slot)->register_handler(handler);
+            return;
+        }
+        if (!handler_slot->is_shared_handler()) {
+            if (handler_slot->type() == HandlerType::SpuriousInterruptHandler) {
+                // FIXME: Add support for spurious interrupts on aarch64
+                VERIFY_NOT_REACHED();
+            }
+            VERIFY(handler_slot->type() == HandlerType::IRQHandler);
+            auto& previous_handler = *handler_slot;
+            handler_slot = nullptr;
+            SharedIRQHandler::initialize(interrupt_number);
+            VERIFY(handler_slot);
```

Comment:
Maybe I'm missing something, but why does calling the SharedIRQHandler::initialize method magically make this handler_slot variable non-null? That is to say, why does that method mess with `s_interrupt_handlers`, when we're in `register_generic_interrupt_handler` here already? I would hope we can refactor the code to only mess with s_interrupt_handlers from the "base class" itself. 

I'm guessing this is just pre-existing design decisions from the x86 impl? if so, no change requested, but it does look odd.

Reply:
It basically ends up calling `register_generic_interrupt_handler` again, but now the `handler_slot` is null, thus it will just assign the newly allocated `SharedIRQHandler` to the slot.

I indeed copied this code from the x86 impl, and I agree that this code should be refactored some time. I think we can design this in such a way that the handlers are stored in the `InterruptManagement` class (or a seperate class), which means we can also move the `register_generic_interrupt_handler` functions to this class.

Reply:
Discussed on Discord: Requires a rework of the code to make it nicer to use. But that can go into another PR.